src/net/discovery/mod.rs

```diff
@@ -0,0 +1,168 @@
+                }
+
+                fn visit_str<E: serde::de::Error>(self, v: &str) -> Result<Self::Value, E> {
+                    let uri = v.parse::<Uri>().map_err(|err| E::custom(err.to_string()))?;
+                    let parts = uri.into_parts();
+                    let scheme = parts.scheme.ok_or_else(|| E::custom("missing scheme"))?;
+                    let authority = parts
+                        .authority
+                        .ok_or_else(|| E::custom("missing authority"))?;
+                    Ok(Origin { scheme, authority })
+                }
+            }
+            deserializer.deserialize_str(UrlVisitor)
+        }
+    }
+
+    #[derive(PartialEq, Eq, Hash, Copy, Clone, Debug)]
+    pub struct PublicKey(x25519_dalek::PublicKey);
+
+    impl Serialize for PublicKey {
+        fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+        where
+            S: Serializer,
+        {
+            let pk_bytes = self.0.to_bytes();
+            let hex_str = hex::encode(pk_bytes);
+            serializer.serialize_str(&hex_str)
+        }
+    }
+
+    impl<'de> Deserialize<'de> for PublicKey {
+        fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+        where
+            D: Deserializer<'de>,
+        {
+            struct PublicKeyVisitor;
+            impl<'de> serde::de::Visitor<'de> for PublicKeyVisitor {
+                type Value = PublicKey;
+
+                fn expecting(&self, formatter: &mut Formatter) -> std::fmt::Result {
+                    formatter.write_str("a valid 32 byte hex string")
+                }
+
+                fn visit_str<E: serde::de::Error>(self, v: &str) -> Result<Self::Value, E> {
+                    let hex_bytes = hex::decode(v).map_err(E::custom)?;
+                    let hex_bytes_array: [u8; 32] =
+                        hex_bytes.try_into().map_err(|v: Vec<u8>| {
+                            E::custom(format!("invalid length for public key: {}", v.len()))
+                        })?;
+                    Ok(PublicKey(hex_bytes_array.into()))
+                }
+            }
+            deserializer.deserialize_str(PublicKeyVisitor)
+        }
+    }
+
+    impl Deref for PublicKey {
+        type Target = x25519_dalek::PublicKey;
+
+        fn deref(&self) -> &Self::Target {
+            &self.0
+        }
+    }
+
+    impl DerefMut for PublicKey {
+        fn deref_mut(&mut self) -> &mut Self::Target {
+            &mut self.0
+        }
+    }
+
+    impl From<[u8; 32]> for PublicKey {
+        fn from(bytes: [u8; 32]) -> Self {
+            PublicKey(x25519_dalek::PublicKey::from(bytes))
+        }
+    }
+
+    #[derive(Clone, Serialize, Deserialize)]
+    pub struct HttpConfig {
+        pub origin: Origin,
+        pub public_key: PublicKey,
+    }
+
+    #[derive(Clone, Serialize, Deserialize)]
+    pub struct PrssConfig {
+        pub public_key: PublicKey,
```

Comment:
should I be using a different type than `x25519_dalek::PublicKey` here?

Reply:
hex vends serde serializers: https://docs.rs/hex/latest/src/hex/serde.rs.html#12

Reply:
ok. should I just be returning a byte array for the public_key then?